### ISSUE TITLE
Supporting multiple API versions

### ISSUE THREAD

**Issue by gravitystorm** [gravitystorm/openstreetmap-website#28]
I'd like to support multiple API versions, so that we can deploy API 0.7 in parallel to API 0.6.

For a start, we need to have a way to have multiple API versions in the codebase (e.g. 0.6 and 0.7), but allow sysadmins to only deploy one version (e.g. 0.6). This will allow us to develop 0.7, iterate, get feedback, change our minds etc all in the master branch, and then allow the sysadmins to choose when to enable 0.7 by just changing a config option. Otherwise, 0.7 support would need to live on a long-running branch as each decision is made and I don't fancy that. The config option can then be used in future to switch off 0.6 in production (e.g. a year later), separately from making any code releases.

The obvious way to do this is to change the `api_version` configuration in application.yml to be an `api_versions` array, currently only containing `["0.6"]`, and reworking the relevant code. The deployed versions can then be e.g. `deployed_versions = API_VERSIONS & ["0.6", "0.7"]` i.e. the intersection of the configuration option and all the versions the codebase knows about. This also allows the tests to work on both versions, even if only one version is available in production.

The next consideration is around the routes and the controllers. It's not possible to do something like this:

```
deployed_versions.each do |version|
    scope "api/#{version}", :defaults => { :api_version => version } do
      get "capabilities" => "api#capabilities"
      get "permissions" => "api#permissions"
      ...
      get "changeset/:id/download" => "changesets#download", :as => :changeset_download, :id => /\d+/
```

since the scope doesn't affect the named routes, and therefore you can't have the same named route appear multiple times. I need to investigate using namespaces instead of scopes here.

We will also need to arrange the controller code, bearing in mind the situations that we want to cover:

* The vast majority of the API will be identical, so we don't want to copy+paste all of the API controller code
* Not all of our code uses views, so the solution can't be only based on changes to the views (this is how some gems work)

There's a few interesting articles that I've found, such as:

* https://paweljw.github.io/2017/07/rails-5.1-api-app-part-3-api-versioning/
* https://yalantis.com/blog/api-versioning-with-ruby-on-rails/
* https://chriskottom.com/blog/2017/04/versioning-a-rails-api/

They all point in roughly the same direction. It remains to be seen whether using a gem to help would be worthwhile, or whether hand-cranking it will be straightforward enough. In particular, for naming things, like controllers, I wish our urls were e.g. "/v6/" instead of "/0.6/"!

**Comment by gravitystorm** [comment]
Something like this will work:

```
OpenStreetMap::Application.routes.draw do
  # Our api versions are decimals, but controllers cannot start with a number
  # or contain punctuation
  def v_string(version)
    "v#{version.delete('.')}"
  end

  # API
  get "api/capabilities" => "api#capabilities"

  scope "api" do
    # For paths common to every version, we use controllers directly in the api/ directory
    ["0.6", "0.7"].each do |version|
      namespace v_string(version), :path => version, :module => :api do
        get "capabilities" => "api#capabilities"
        get "permissions" => "api#permissions"
....
    # Different controllers for different versions
    ["0.6", "0.7"].each do |version|
      namespace v_string(version), :path => version, :module => "api/#{v_string(version)}" do
        get "node/create" => "nodes#create"
      end
    end
```

This gives routes like

```
                 v06_changeset_create PUT      /api/0.6/changeset/create(.:format)                                                      api/changesets#create
                 v07_changeset_create PUT      /api/0.7/changeset/create(.:format)                                                      api/changesets#create
```
... for stuff that hasn't changed, and 

```
                      v06_node_create GET      /api/0.6/node/create(.:format)                                                           api/v06/nodes#create
                      v07_node_create GET      /api/0.7/node/create(.:format)                                                           api/v07/nodes#create
```

... for code that has changed. The corresponding classes would be e.g. `Api::ChangesetsController < ApiController` and `Api::v07::NodesController < Api::NodesController` etc.

First we need to refactor the controllers, but now we have an end-goal nailed down. See #27

**Comment by gravitystorm** [comment]
I've rebased and reworked this today onto https://github.com/gravitystorm/openstreetmap-website/tree/zero_point_seven_mk3

It's hard to split this into smaller steps, since I need to make sure the end result is completely workable for all the different things we want to do in 0.7 (remove some features, rework other features, etc). So I want to push on a bit into refactoring all the routes and see what happens. But so far so good.

The biggest problem is the tests. I believe we need to test all the versions. But this involves adding a whole lot of loops (`all_api_versions.each do |version|`) which destroys the git history and makes the diffs unreadable. So I plan to do this first, before refactoring any API methods, so that the history is a little clearer. But a way to avoid indentation would be great.

Additionally, I haven't figured out a nicer solution yet for:

```diff
- get :show
+ get :show, :params => { :api_version => version }
```
The above requires a ton of typing and reading, so it's not ideal. Maybe some refactoring will present itself later.

**Comment by gravitystorm** [comment]
> Additionally, I haven't figured out a nicer solution yet

I spent a few hours on this today, learning a lot about blocks, procs, and bindings. The following pretty much works:

```ruby
    def with_all_api_versions
      ["0.6", "0.7"].each do |version|
        class_eval <<-RUBY, __FILE__, __LINE__ + 1
         def api_get(x)
           get x, :params => { :api_version => #{version} }
         end
        RUBY
        yield
      end
    end
```

That allows you to write `api_get :show` in the test, and it calls `get :show, :params => {...etc...}` in the background. That is, you don't need to specify the version in every get call, since the method is redefined each time we yield to the tests.

However, there's two things that count against this:
* First, it will require a lot of complex definitions for `api_put`, `api_post` etc
* Second, it seems most tests need the `version` variable anyway. So a simpler approach would just be `yield(version)` and `api_get(version) :show`, which avoids the metaprogramming.

I think for now I'll just stick to a straightforward refactoring and see what the comments are on that.
